### ISSUE TITLE
Blacklisted packages not upgraded but not in pkgs_kept_back

### ISSUE THREAD

**Issue by lafrech** [mvo5/unattended-upgrades#52]
Hi.

I set a few package in blacklist:

```python
Unattended-Upgrade::Package-Blacklist {
        "vim";
        "libc6";
        "libc6-dev";
        "libc6-i686";
        "dovecot";
        "roundcube";
        "postgresql";
        "mysql-server";
};
```

Some of these currently have upgrades available.

They are hold back as expected.

However, I expected them to appear in the "Packages with upgradable origin but kept back" part of the mail, and they don't.

The only package that appears in the mail is `libc-dev-bin`.

I can't tell whether this is an issue or I'm just misunderstanding something.

I thought the point of this section of the mail was to inform the user that some packages could be upgraded. This does not seem to be the case. If I don't launch `apt-get`/`aptitude` myself, I won't be notified.

I am indeed notified about the libc update (the mail mentions `libc-dev-bin`) but not about postgres and vim updates, for instance.

(I was using Debian's stable version (0.83.3.2+deb8u1). I upgraded to latest Debian package (0.93.1+nmu1) and the behavior is the same.)

**Comment by GhostLyrics** [comment]
The section in the mail is to be read as "apt tries to upgrade the package but it was a failure.( Or it could not be upgraded due to dependency issues)". 

It is not what you expect. What you expect is "apt was not allowed to upgrade this package."

**Comment by lafrech** [comment]
Thanks @GhostLyrics, it makes more sense to me, now.

- `libc-dev-bin` can't upgrade because it `replaces` a version of `libc6-dev` that itself can not be updated due to blacklisting, hence the "failure or dependency issue" and the message.

- The other packages can't be replaced because they are blacklisted more directly.

Wouldn't it be more consistent to notify the admin in both cases?

AFAIU, the point of unattended-upgrades is to avoid launching apt-get/aptitude regularly and let it manage upgrades, either updating automatically, either notifying the admin. This purpose gets defeated if it silently ignores blacklisted upgrades.

There are other tools to be notified but using them would be redundant and the admin would be notified for every upgradable package, even the 99% that are automatically upgraded.

Would it make sense to notify about all upgradable but not upgraded packages?

- failures
- dependency issues
- blacklisted
- origin not allowed

This could be done by adding them to `pkgs_kept_back`, or if we don't want to change the "package on hold" behavior, by adding another category, but I don't really see the point of a distinction between the two categories, so merging them in existing `pkgs_kept_back` is my favorite option.

**Comment by jayaich** [comment]
> AFAIU, the point of unattended-upgrades is to avoid launching apt-get/aptitude regularly and let it manage upgrades, either updating automatically, either notifying the admin. This purpose gets defeated if it silently ignores blacklisted upgrades.

I very much agree. If a package is blacklisted, but has no held back dependencies, then you would receive no notification that the package has an update available. I want unattended upgrades for basic commands such as 'ls', 'date' etc, but also want updates to the kernel and C library held back (blacklisted). For those packages I will schedule a patch and reboot, and manually run apt-get. However, I need to be notified if updates for those packages are available.

**Comment by rbalint** [comment]
This can be closed now since it was fixed in PR #63.

**Comment by lafrech** [comment]
Yes, thanks @jayaich for the PR.
